swap/src/protocol/alice/refund.rs

```diff
@@ -0,0 +1,123 @@
+        match state {
+
+            // In case no XMR has been locked, move to Safely Aborted
+            AliceState::Started { .. }
+            | AliceState::BtcLocked { .. } => bail!(Error::NoXmrLocked(state)),
+
+            // Refund potentially possible (no knowledge of cancel transaction)
+            AliceState::XmrLockTransactionSent { monero_wallet_restore_blockheight, transfer_proof, state3, }
+            | AliceState::XmrLocked { monero_wallet_restore_blockheight, transfer_proof, state3 }
+            | AliceState::XmrLockTransferProofSent { monero_wallet_restore_blockheight, transfer_proof, state3 }
+            | AliceState::EncSigLearned { monero_wallet_restore_blockheight, transfer_proof, state3, .. }
+            | AliceState::CancelTimelockExpired { monero_wallet_restore_blockheight, transfer_proof, state3 }
+
+            // Refund possible due to cancel transaction already being published
+            | AliceState::BtcCancelled { monero_wallet_restore_blockheight, transfer_proof, state3 }
+            | AliceState::BtcRefunded { monero_wallet_restore_blockheight, transfer_proof, state3, .. }
+            | AliceState::BtcPunishable { monero_wallet_restore_blockheight, transfer_proof, state3, .. } => {
+                (monero_wallet_restore_blockheight, transfer_proof, state3)
+            }
+
+            // Alice already in final state
+            AliceState::BtcRedeemed
+            | AliceState::XmrRefunded
+            | AliceState::BtcPunished
+            | AliceState::SafelyAborted => bail!(Error::SwapNotRefundable(state)),
+        }
+    } else {
+        match state {
+            AliceState::Started { .. } | AliceState::BtcLocked { .. } => {
+                bail!(Error::NoXmrLocked(state))
+            }
+
+            AliceState::BtcCancelled {
+                monero_wallet_restore_blockheight,
+                transfer_proof,
+                state3,
+            }
+            | AliceState::BtcRefunded {
+                monero_wallet_restore_blockheight,
+                transfer_proof,
+                state3,
+                ..
+            }
+            | AliceState::BtcPunishable {
+                monero_wallet_restore_blockheight,
+                transfer_proof,
+                state3,
+                ..
+            } => (monero_wallet_restore_blockheight, transfer_proof, state3),
+
+            AliceState::BtcRedeemed
+            | AliceState::XmrRefunded
+            | AliceState::BtcPunished
+            | AliceState::SafelyAborted => bail!(Error::SwapNotRefundable(state)),
+
+            _ => return Ok(Err(Error::SwapNotCancelled)),
+        }
+    };
+
+    tracing::info!(%swap_id, "Trying to manually refund swap");
+
+    let spend_key = if let Ok(published_refund_tx) =
+        state3.check_for_tx_refund(bitcoin_wallet.as_ref()).await
+    {
+        tracing::debug!(%swap_id, "Bitcoin refund transaction found, extracting key to refund Monero");
```

Comment:
Print the txid here?

Reply:
Also, the user passed you the swap ID, seems redundant to include it in the log here.